### ISSUE TITLE
Add settings to block various types of requests

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#2626]
This is sort of the first step torwards uMatrix-like functionality: Having a way to block various requests based from data in [QWebEngineUrlRequestInfo](https://doc.qt.io/qt-5/qwebengineurlrequestinfo.html) (per-domain after #27 of course, which is pretty much uMatrix minus the UI).

The main interesting thing there is [ResourceType](https://doc.qt.io/qt-5/qwebengineurlrequestinfo.html#ResourceType-enum), with the following types:

- MainFrame	(Top level page)
  - Probably no point in allowing to block this
- SubFrame (Frame or iframe)
  - Should be blockable
  - Blockable in uMatrix
  - Might be blockable in QtWebKit too via [acceptNavigationRequest](http://doc.qt.io/archives/qt-5.5/qwebpage.html#acceptNavigationRequest)
- Stylesheet (A CSS stylesheet)
  - Should be blockable
  - Blockable in uMatrix
  - Probably not blockable in QtWebKit (maybe via mimetype?)
- Script (An external script)
  - Already handled via `content -> allow-javascript`.
  - If they still occur with the setting off, might be good to block them
  - Blockable in uMatrix
- Image (An image (JPG, GIF, PNG, and so on))
  - Already handled via `content -> allow-images`.
  - If they still occur with the setting off, might be good to block them
  - Blockable in uMatrix
- FontResource (A font)
  - Should be blockable if those are web fonts
  - Not blockable in uMatrix
  - Probably not blockable in QtWebKit (maybe via mimetype?)
- SubResource (An "other" subresource)
  - Unclear when it happens... Probably fine to leave it right now.
- Object (An object (or embed) tag for a plugin or a resource that a plugin requested)
  - Should maybe be blockable?
  - Not blockable in uMatrix
  - Probably not blockable in QtWebKit (maybe via mimetype?)
- Media (A media resource)
  - Should be blockable
  - Blockable in uMatrix
  - Maybe blockable in QtWebKit via mimetype?
  - QtWebKit-NG has [`QWebSettings::MediaEnabled`](https://github.com/annulen/webkit/commit/856fd172995f9) - we should be able to use the int value without PyQt support
  - Doesn't seem to block YouTube though, as that uses XHR requests...
  - See https://github.com/annulen/webkit/issues/357 for click-to-play on QtWebKit; QtWebEngine still needs an issue for this
- Worker (The main resource of a dedicated worker)
  - Should maybe be blockable?
  - Not blockable in uMatrix
  - Probably not blockable in QtWebKit
- SharedWorker (The main resource of a shared worker)
  - Should maybe be blockable?
  - Not blockable in uMatrix
  - Probably not blockable in QtWebKit
- Prefetch (An explicitly requested prefetch)
  - Unclear when it happens...
  - Maybe this would allow to implement `network -> dns-prefetch` for QtWebEngine?
  - Not blockable in uMatrix
- Favicon
  - Already handled via `content -> allow-images`.
  - If they still occur with the setting off, might be good to block them
  - Not blockable in uMatrix
- Xhr (An XMLHttpRequest)
  - Should be blockable
  - Blockable in uMatrix
  - Probably not blockable in QtWebKit
- Ping (A ping request for <a ping>)
  - Already handled via `content -> hyperlink-auditing`.
  - If they still occur with the setting off, might be good to block them
  - Blockable in uMatrix
- ServiceWorker (The main resource of a service worker)
  - Should maybe be blockable?
  - Not blockable in uMatrix
  - Probably not blockable in QtWebKit
- CspReport (A report of Content Security Policy (CSP) violations)
  - Should maybe be blockable?
  - Not blockable in uMatrix
  - Probably not blockable in QtWebKit
- PluginResource (A resource requested by a plugin)
  - Should maybe be blockable?
  - Not blockable in uMatrix
  - Probably not blockable in QtWebKit
- Unknown / Other
  - Should be blockable
  - Blockable in uMatrix
  - Probably not blockable in QtWebKit

**Comment by lahwaacz** [comment]
Loading of remote fonts can be disabled with `--qt-flag disable-remote-fonts`. Is blocking `FontResource` fully equivalent?

**Comment by The-Compiler** [comment]
From a quick look at Chromium's code, they seem equivalent - but it's a hell of lasagna code (too many layers), so I can't say I'm 100% confident

**Comment by paulclarkaranz** [comment]
Does anyone have an idea of what the UI for this would look like? I get how the blocking would be handled mostly, but I'm not sure of a good way of letting the user configure it. Or would it just be `:config-source`?

**Comment by toofar** [comment]
Does anyone have an idea of what the UI for this would look like? I get how the blocking would be implemented, mostly. But how best to let the user configure it? Or would it just be `:config-source`?

**Comment by The-Compiler** [comment]
For this issue and added from-url support, yeah, it'd be something like `config.set('content.block.xhr`, True, url='http://example.com/', from_url='http://www.example.org/')` and `:set --url http://example.com/ --from-url http://example.org/ content.block.xhr` or so.

With #28 there should be a proper UI for the same, similar like [uMatrix](https://addons.mozilla.org/en-US/firefox/addon/umatrix/) (plus probably `:matrix-*` commands with completion or so).

**Comment by jakecoble** [comment]
Would `from_url` be an extension of the per-domain settings in this case, thereby allowing something like the following?

```
with config.from_url('*://example.com') as fu:
    fu.content.block.xhr = true
```

**Comment by The-Compiler** [comment]
@jakecoble Yeah, though you'd probably want to specify both `url` and `from_url`, see my comment above.

**Comment by def14nt** [comment]
Can I do anything to help? I'd love to speed this up, as it's actually the only reason *not* to use qutebrowser. `(Firefox|Chrome)` with a vim extension isn't the same.

**Comment by The-Compiler** [comment]
@def14nt Not really. I'm currently working on a [plugin API](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2018-September/000051.html) which also should make this kind of thing easier (inside a plugin) in the future though.

**Comment by def14nt** [comment]
@The-Compiler Nice! Hopefully people will adopt the plugin ecosystem for qutebrowser.

**Comment by The-Compiler** [comment]
I've had some discussions in IRC about this - I think this is actually something which would provide a lot of value even without the `from_url` change.

To avoid having to add some ~20 new settings, it should probably be a `content.blocked_request_types` or so instead, which takes a list - then, you could e.g. do `:set -u example.org content.blocked_request_types '["xhr"]'` or so.

Finally, I don't think this needs QtWebKit support anymore at this point - let's just make it a QtWebEngine-only setting.

With all those constraints, this would be a good first issue for a contribution.

**Comment by toofar** [comment]
(Assuming I'm not missing something, although I haven't been following that.) Should this only handle things not covered by the brave adblocker or is overlap okay?

**Comment by The-Compiler** [comment]
Ah, that's a good point - like you say, some of those types can be [set in adblock rules](https://adblockplus.org/filter-cheatsheet#filter-options) as well.

Perhaps this is actually a better replacement for the `from_url` idea I had mentioned earlier, I'll need to think some more about that. I still think there's value in having a simple "block those kinds of requests on the given website" toggle. If we add such a toggle anyways (and as long as it's a single option taking a list of types),

I don't think there's a good reason to only support a subset of request types - as usual, we let users shoot into their feet if they insist on doing so :wink:

At least that's my take - while only tangentially related to the adblocker, I wonder if @ArniDagur has an opinion on this one?
